group/assignment5/AlphabetFromDictionary.cpp

```diff
@@ -0,0 +1,94 @@
+#include "AlphabetFromDictionary.hpp"
+
+void AlphabetFromDictionary::setDictionary(vector<string> dictionary) {
+    this->dictionary = dictionary;
+}
+
+void AlphabetFromDictionary::initializeClassVars() {
+    vector<char> alphabetCharacters = findAlphabetCharacters();
+    this->charPriorityGraph.clear();
+    this->numOfPriorCharacters.clear();
+    for (auto c : alphabetCharacters) {
+        this->charPriorityGraph[c] = set<char>();
+        this->numOfPriorCharacters[c] = 0;
+    }
+}
+
+bool AlphabetFromDictionary::findAlphabet(vector<char> &alphabet) {
+    this->initializeClassVars();
+    this->createCharPriorityGraph();
+
+    // find topological sort of characters graph to find alphabet
+    return topologicalSort(alphabet);
+}
+
+void AlphabetFromDictionary::updateNumOfPriorCharacters(const char origin) {
+    const set<char> followingChars = this->charPriorityGraph.at(origin);
+    for (char c : followingChars) {
+        this->numOfPriorCharacters[c]--;
+    }
+    this->numOfPriorCharacters.erase(origin);
+}
+
+bool AlphabetFromDictionary::findNextOrigin(char &origin) {
+    for (auto it : this->numOfPriorCharacters) {
+        if (it.second == 0) {
+            origin = it.first;
+            return true;
+        }
+    }
+    // no origin -> circle in graph -> dictionary is inconsistent
+    return false;
+}
+
+bool AlphabetFromDictionary::topologicalSort(vector<char> &alphabet) {

```

Comment:
Solution finds topological sorting but it's not the algorithm of topological sort, I think complexity will improve if we use algorithm because it goes in each node once whereas this solution searches through all chars to find next char updates their priors and erases all the chars. Or we can just sort the map numOfPriorCharacters by second element. For optional #2 we can detect cyrcle during DFS or if we sort the map then we can check if consecutive numbers does not differ by 1. But sorting takes O(nlogn), so first option is better

Reply:
I agree, let's use the known solution and use DFS for challenge 2